File: 프로그래머스/lv2/76502. 괄호 회전하기/괄호 회전하기.js
function solution(s) {
    let count = 0;
    const n = s.length;
    const pair = {'(' : ')', '{' : '}', '[' : ']'};

    for (let i  = 1; i <= n; i++) {
        const stack = [];
        
        if (!pair[s[0]]) { // 첫번째부터 닫는 괄호라면 얼리 리턴
            console.log('얼리 리턴', pair[s[0]])
            s = s.substr(1) + s[0];
            continue;
        }; 
        
        for (const bracket of s) {
            if (pair[bracket]) { // 여는 괄호면 스택에 넣고,
                stack.push(bracket);
                continue;
            }
            
            const top = stack[stack.length - 1]
            if (pair[top] === bracket) { // 닫는 괄호면, 스택의 맨 위 요소와 짝이 맞는다면
                stack.pop(); // pop해준다.
                continue;
            }
        }
        
        if (stack.length === 0) count++; // 올바른 괄호라면 카운트 + 1
        s = s.substr(1) + s[0];
    }
    
    return count;
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.py
def isRightBracketString(s):
    stack = []
    
    for bracket in s:
        if bracket in ['[', '{', '(']:
            stack.append(bracket)
        elif bracket in [']', '}', ')']:
            if not stack:
                return False
            
            top = stack[-1]
            if (top == '[' and bracket == ']') or \
               (top == '{' and bracket == '}') or \
               (top == '(' and bracket == ')'):
                stack.pop() 
        else:
            return False
    return len(stack) == 0

def solution (s):
    count = 0
    for i in range(len(s)):
        if isRightBracketString(s):
            count += 1
        s = s[1:] + s[0]
            
    return count